dlib/matrix/matrix_fft.h

```diff
@@ -9,837 +9,142 @@
-    */
-#if 0
-#ifdef DLIB_USE_FFTW
-
-    template <long NR, long NC, typename MM, typename L>
-    matrix<std::complex<double>,NR,NC,MM,L> call_fftw_fft(
-        const matrix<std::complex<double>,NR,NC,MM,L>& data
-    )
+#else
+    
+// ----------------------------------------------------------------------------------------
+    
+    template <typename EXP>
+    matrix<typename EXP::type> fft (const matrix_exp<EXP>& data)
     {
-        // make sure requires clause is not broken
-        DLIB_CASSERT(is_power_of_two(data.nr()) && is_power_of_two(data.nc()),
-            "\t matrix fft(data)"
-            << "\n\t The number of rows and columns must be powers of two."
-            << "\n\t data.nr(): "<< data.nr()
-            << "\n\t data.nc(): "<< data.nc()
-            << "\n\t is_power_of_two(data.nr()): " << is_power_of_two(data.nr())
-            << "\n\t is_power_of_two(data.nc()): " << is_power_of_two(data.nc())
-            );
-
-        if (data.size() == 0)
-            return data;
-
-        matrix<std::complex<double>,NR,NC,MM,L> m2(data.nr(),data.nc());
-        fftw_complex *in, *out;
-        fftw_plan p;
-        in = (fftw_complex*)&data(0,0);
-        out = (fftw_complex*)&m2(0,0);
-        if (data.nr() == 1 || data.nc() == 1)
-            p = fftw_plan_dft_1d(data.size(), in, out, FFTW_FORWARD, FFTW_ESTIMATE);
-        else
-            p = fftw_plan_dft_2d(data.nr(), data.nc(), in, out, FFTW_FORWARD, FFTW_ESTIMATE);
-        fftw_execute(p); 
-        fftw_destroy_plan(p);
-        return m2;
+        // You have to give a complex matrix
+        COMPILE_TIME_ASSERT(is_complex<typename EXP::type>::value);
+        matrix<typename EXP::type> eval(data); //potentially 2 copies: 1 for evaluating the matrix expression and 1 for doing the out-of-place FFT. hmm...
+        kiss_fft({(int)eval.nr(), (int)eval.nc()}, &eval(0,0), &eval(0,0), false);
+        return eval;
     }
 
-    template <long NR, long NC, typename MM, typename L>
-    matrix<std::complex<double>,NR,NC,MM,L> call_fftw_ifft(
-        const matrix<std::complex<double>,NR,NC,MM,L>& data
-    )
+// ----------------------------------------------------------------------------------------
+    
+    template <typename EXP>
+    matrix<typename EXP::type> ifft (const matrix_exp<EXP>& data)
     {
-        // make sure requires clause is not broken
-        DLIB_CASSERT(is_power_of_two(data.nr()) && is_power_of_two(data.nc()),
-            "\t matrix ifft(data)"
-            << "\n\t The number of rows and columns must be powers of two."
-            << "\n\t data.nr(): "<< data.nr()
-            << "\n\t data.nc(): "<< data.nc()
-            << "\n\t is_power_of_two(data.nr()): " << is_power_of_two(data.nr())
-            << "\n\t is_power_of_two(data.nc()): " << is_power_of_two(data.nc())
-            );
-
+        // You have to give a complex matrix
+        COMPILE_TIME_ASSERT(is_complex<typename EXP::type>::value);
```

Comment:
This can use `static_assert` now.  `COMPILE_TIME_ASSERT` is a pre-c++11 workaround.